core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1067,13 +1130,80 @@ class KafkaController(val config: KafkaConfig,
     (topicsToBeDeleted, topicsIneligibleForDeletion)
   }
 
-  private def updateLeaderAndIsrCache(partitions: Seq[TopicPartition] = controllerContext.allPartitions.toSeq): Unit = {
-    val leaderIsrAndControllerEpochs = zkClient.getTopicPartitionStates(partitions)
+  // sequential version:  very slow for clusters with many topic-partitions (e.g., > 50000)
+  private def updateLeaderAndIsrCache(partitions: Seq[TopicPartition]): Unit = {
+    val zkPartitionRecurseStartMs = time.milliseconds()
+    val leaderIsrAndControllerEpochs = zkClient.getTopicPartitionStates(partitions)  // 98.4% of method's time
+    info(s"partition-level init (sequential): finished recursing ${partitions.size} partitions in ZK ${timing(zkPartitionRecurseStartMs, -1)}")
     leaderIsrAndControllerEpochs.forKeyValue { (partition, leaderIsrAndControllerEpoch) =>
       controllerContext.putPartitionLeadershipInfo(partition, leaderIsrAndControllerEpoch)
     }
   }
 
+  // parallel version:  read topic-partitions info from ZK in multiple threads
+  private def updateLeaderAndIsrCacheParallel(partitions: Set[TopicPartition]): Unit = {
+    if (partitions.isEmpty)
+      return
+
+    val leaderIsrAndControllerEpochs: mutable.Map[TopicPartition, LeaderIsrAndControllerEpoch] = mutable.Map.empty
+
+    // (1) split up partitions set into (at most) liNumControllerInitThreads batches
+    //     The formula below ensures that, if the partition count isn't an exact multiple of the thread count,
+    //     we simply end up with a "short stack" for one of the threads--and that only in the worst case, i.e.,
+    //     where the zip implementation is naive in its bin-packing--rather than more batches than threads,
+    //     which would require one thread to run two batches sequentially and roughly double the overall time.
+    //     (And yes, Scala's zip implementation IS one of the stupider ones:  the ideal split for 13 partitions
```

Comment:
The described behavior is caused by Scala's grouped implementation, instead of zip. Right?

Reply:
Oops, thanks for catching that.